src/js/player.js

```diff
@@ -1177,6 +1178,109 @@ class Player extends Component {
     }
   }
 
+  /**
+   * find the mime type of a given source string url if possible
+   *
+   * @param {string} src
+   *        The source string
+   *
+   * @return {string}
+   *         The type that was found
+   *
+   */
+  findMimeType_(src) {
+    if (!src) {
+      return '';
+    }
+
+    // 1. check for the type in the `source` cache
+    if (this.cache_.source.src === src && this.cache_.source.type) {
+      return this.cache_.source.type;
+    }
+
+    // 2. see if we have this source in our `currentSources` cache
+    const matchingSources = this.cache_.sources.filter((s) => s.src && s.src === src);
+
+    if (matchingSources.length) {
+      return matchingSources[0].type;
+    }
+
+    // 3. look for the src url in source elements and use the type there
+    const sources = this.$$('source');
+
+    for (let i = 0; i < sources.length; i++) {
+      const s = sources[i];
+
+      if (s.type && s.src && s.src === src) {
+        return s.type;
+      }
+    }
+
+    // 4. finally fallback to our list of mime types based on src url extension
+    return getMimeType(src);
+  }
+
+  /**
+   * Update the internal source caches so that we return the correct source from
+   * `src()`, `currentSource()`, and `currentSources()`.
+   *
+   * > Note: `currentSources` will not be updated if the source that is passed in exists
+   *         in the current `currentSources` cache.
+   *
+   *
+   * @param {Tech~SourceObject} srcObj
+   *        A string or object source to update our caches to.
+   */
+  updateSourceCaches_(srcObj = '') {
+    let src = srcObj;
+    let type = '';
+
+    if (typeof src !== 'string') {
+      src = srcObj.src;
+      type = srcObj.type;
+    }
+    // make sure all the caches are set to default values
+    // to prevent null checking
+    this.cache_.source = this.cache_.source || {};
+    this.cache_.sources = this.cache_.sources || [];
+
+    // try to get the type of the src that was passed in
+    if (src && !type) {
+      type = this.findMimeType_(src);
+    }
+
+    // update `currentSource` cache always
+    this.cache_.source = {src, type};
+
+    const matchingSources = this.cache_.sources.filter((s) => s.src && s.src === src);
+    const sourceElSources = [];
+    const sourceEls = this.$$('source');
+    const matchingSourceEls = [];
+
+    for (let i = 0; i < sourceEls.length; i++) {
+      const sourceObj = {src: sourceEls[i].src, type: sourceEls[i].type};
```

Comment:
What if the `source` has no `type` attribute? It doesn't seem to be required.

Reply:
It should return undefined or empty string in that case and we may want to omit it in that case. I will test this.